File: API/app/services/todo_service.py
from datetime import date, datetime, timezone
from sqlalchemy.orm import Session, joinedload
from app.models.todo import RecurringTodo, TimelineTodo, DailyList, DailyTask
from app.schemas.todo import (
    RecurringTodoCreate, RecurringTodoUpdate,
    TimelineTodoCreate, TimelineTodoUpdate,
    DailyTaskCreate,
)


# ── Recurring ────────────────────────────────────────────────

def get_recurring(db: Session) -> list[RecurringTodo]:
    return db.query(RecurringTodo).order_by(RecurringTodo.title).all()


def create_recurring(db: Session, data: RecurringTodoCreate) -> RecurringTodo:
    todo = RecurringTodo(**data.model_dump())
    db.add(todo)
    db.commit()
    db.refresh(todo)
    return todo


def update_recurring(db: Session, todo_id: int, data: RecurringTodoUpdate) -> RecurringTodo | None:
    todo = db.query(RecurringTodo).filter(RecurringTodo.id == todo_id).first()
    if not todo:
        return None
    for k, v in data.model_dump().items():
        setattr(todo, k, v)
    todo.updated_at = datetime.now(timezone.utc)
    db.commit()
    db.refresh(todo)
    return todo


def toggle_recurring(db: Session, todo_id: int) -> RecurringTodo | None:
    todo = db.query(RecurringTodo).filter(RecurringTodo.id == todo_id).first()
    if not todo:
        return None
    todo.is_active = not todo.is_active
    todo.updated_at = datetime.now(timezone.utc)
    db.commit()
    db.refresh(todo)
    return todo


def delete_recurring(db: Session, todo_id: int) -> bool:
    todo = db.query(RecurringTodo).filter(RecurringTodo.id == todo_id).first()
    if not todo:
        return False
    db.delete(todo)
    db.commit()
    return True


# ── Timeline ─────────────────────────────────────────────────

def get_timeline(db: Session, status: str | None = None) -> list[TimelineTodo]:
    q = db.query(TimelineTodo)
    if status:
        q = q.filter(TimelineTodo.status == status)
    return q.order_by(TimelineTodo.due_date.asc()).all()


def create_timeline(db: Session, data: TimelineTodoCreate) -> TimelineTodo:
    todo = TimelineTodo(**data.model_dump())
    db.add(todo)
    db.commit()
    db.refresh(todo)
    return todo


def update_timeline(db: Session, todo_id: int, data: TimelineTodoUpdate) -> TimelineTodo | None:
    todo = db.query(TimelineTodo).filter(TimelineTodo.id == todo_id).first()
    if not todo:
        return None
    for k, v in data.model_dump().items():
        setattr(todo, k, v)
    todo.updated_at = datetime.now(timezone.utc)
    db.commit()
    db.refresh(todo)
    return todo


def patch_timeline_status(db: Session, todo_id: int, status: str) -> TimelineTodo | None:
    todo = db.query(TimelineTodo).filter(TimelineTodo.id == todo_id).first()
    if not todo:
        return None
    todo.status = status
    todo.updated_at = datetime.now(timezone.utc)
    db.commit()
    db.refresh(todo)
    return todo


def delete_timeline(db: Session, todo_id: int) -> bool:
    todo = db.query(TimelineTodo).filter(TimelineTodo.id == todo_id).first()
    if not todo:
        return False
    db.delete(todo)
    db.commit()
    return True


# ── Daily ─────────────────────────────────────────────────────

def get_daily_list(db: Session, list_date: date) -> DailyList | None:
    return (
        db.query(DailyList)
        .options(joinedload(DailyList.tasks))
        .filter(DailyList.date == list_date)
        .first()
    )


def get_or_create_daily_list(db: Session, list_date: date) -> DailyList:
    existing = get_daily_list(db, list_date)
    if existing:
        return existing
    dl = DailyList(date=list_date)
    db.add(dl)
    db.commit()
    db.refresh(dl)
    return db.query(DailyList).options(joinedload(DailyList.tasks)).filter(DailyList.id == dl.id).first()


def get_all_daily_lists(db: Session) -> list[DailyList]:
    return db.query(DailyList).order_by(DailyList.date.desc()).all()


def add_daily_task(db: Session, list_id: int, data: DailyTaskCreate) -> DailyTask:
    task = DailyTask(list_id=list_id, **data.model_dump())
    db.add(task)
    db.commit()
    db.refresh(task)
    return task


def toggle_daily_task(db: Session, task_id: int) -> DailyTask | None:
    task = db.query(DailyTask).filter(DailyTask.id == task_id).first()
    if not task:
        return None
    task.is_done = not task.is_done
    task.updated_at = datetime.now(timezone.utc)
    db.commit()
    db.refresh(task)
    return task


def delete_daily_task(db: Session, task_id: int) -> bool:
    task = db.query(DailyTask).filter(DailyTask.id == task_id).first()
    if not task:
        return False
    db.delete(task)
    db.commit()
    return True

File: UI/src/app/core/services/todo.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import {
  RecurringTodo, RecurringTodoPayload,
  TimelineTodo, TimelineTodoPayload,
  DailyList, DailyTask,
} from '../../shared/models/todo.model';
import { environment } from '../../../environments/environment';

@Injectable({ providedIn: 'root' })
export class TodoService {
  private base = `${environment.apiBaseUrl}/todos`;

  constructor(private http: HttpClient) {}

  // ── Recurring
  getRecurring(): Observable<RecurringTodo[]> {
    return this.http.get<RecurringTodo[]>(`${this.base}/recurring/`);
  }
  createRecurring(data: RecurringTodoPayload): Observable<RecurringTodo> {
    return this.http.post<RecurringTodo>(`${this.base}/recurring/`, data);
  }
  updateRecurring(id: number, data: RecurringTodoPayload): Observable<RecurringTodo> {
    return this.http.put<RecurringTodo>(`${this.base}/recurring/${id}`, data);
  }
  toggleRecurring(id: number): Observable<RecurringTodo> {
    return this.http.patch<RecurringTodo>(`${this.base}/recurring/${id}/toggle`, {});
  }
  deleteRecurring(id: number): Observable<void> {
    return this.http.delete<void>(`${this.base}/recurring/${id}`);
  }

  // ── Timeline
  getTimeline(status?: string): Observable<TimelineTodo[]> {
    let params = new HttpParams();
    if (status) params = params.set('status', status);
    return this.http.get<TimelineTodo[]>(`${this.base}/timeline/`, { params });
  }
  createTimeline(data: TimelineTodoPayload): Observable<TimelineTodo> {
    return this.http.post<TimelineTodo>(`${this.base}/timeline/`, data);
  }
  updateTimeline(id: number, data: TimelineTodoPayload): Observable<TimelineTodo> {
    return this.http.put<TimelineTodo>(`${this.base}/timeline/${id}`, data);
  }
  patchTimelineStatus(id: number, status: string): Observable<TimelineTodo> {
    return this.http.patch<TimelineTodo>(`${this.base}/timeline/${id}/status`, { status });
  }
  deleteTimeline(id: number): Observable<void> {
    return this.http.delete<void>(`${this.base}/timeline/${id}`);
  }

  // ── Daily
  getDailyByDate(date: string): Observable<DailyList | null> {
    return this.http.get<DailyList | null>(`${this.base}/daily/by-date`, { params: { date } });
  }
  createDailyList(date: string): Observable<DailyList> {
    return this.http.post<DailyList>(`${this.base}/daily/by-date`, {}, { params: { date } });
  }
  addDailyTask(listId: number, title: string, sortOrder = 0): Observable<DailyTask> {
    return this.http.post<DailyTask>(`${this.base}/daily/${listId}/tasks`, { title, sort_order: sortOrder });
  }
  toggleDailyTask(taskId: number): Observable<DailyTask> {
    return this.http.patch<DailyTask>(`${this.base}/daily/tasks/${taskId}/toggle`, {});
  }
  deleteDailyTask(taskId: number): Observable<void> {
    return this.http.delete<void>(`${this.base}/daily/tasks/${taskId}`);
  }
}
